### PR TITLE
Implement PushTarget preference

### PR DESCRIPTION
**Pull request by johnmcollier**
**What type of PR is this?**

/kind feature

**What does does this PR do / why we need it**:

This PR implements the `PushTarget` setting as described in https://github.com/openshift/odo/issues/2577.

The push target for odo can be set in two ways:
   - `odo preference set pushtarget <target>`, which will add the following to the preference.yaml file:
      ```
      OdoSettings:
        PushTarget: docker
      ```
      or
       ```
      OdoSettings:
        PushTarget: kube
      ```
   - Setting the `ODO_PUSHTARGET=<target>` environment variable, which will override the value that's set in the preference.yaml file
    

As this functionality is currently a no-op (pending the implementation of odo push for docker), I've left it out of any of the descriptions in `odo preference` or `odo preference set/unset` to avoid confusing users. 

This PR adds the following:
- New `PushTarget` preference
   - Valid values are `kube` and `docker`.
   - PreferenceInfo.GetPushTarget() returns `kube` if no field is set in the preferences
- New `pushtarget` package in `pkg/odo/util` to help determining if the push target is set to docker (via `IsPushTargetDocker()`)
   - Checks the environment variable first, and if the environment variable is not set, it will check the preference.yaml field for `PushTarget`
- Unit tests
   - Unit tests for the new preference `PushTarget`
      - Tests both valid and invalid values
   - Unit tests for the new `pushtarget` package
      - Verifies that `IsPushTargetDocker` returns the proper values based on the values of `ODO_PUSHTARGET` and the `PushTarget` preference

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/2577

**How to test changes / Special notes to the reviewer**:
1. Build my branch and build odo
2. Run `make test` to verify the tests I added in the `preferences` and `pushtarget` packages pass
3. Run `odo preference set pushtarget <target>` to set the push target in the preferences.yaml file (valid values are docker and kube). Invalid values should fail.
4. Run `odo preference unset pushtarget` to verify that the setting gets removed from preference.yaml

### CODE REVIEW

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	// and will override the setting in the preferences file.
+
+	OdoPushTarget = "ODO_PUSHTARGET"
+)
+
+// IsPushTargetDocker checks if the push target preference has been set to docker
+func IsPushTargetDocker() bool {
+
+	// ODO's push target can be told to use docker by:
+	//		- setting an env variable "ODO_PUSHTARGET=docker" or
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget.go]: if the user has set both but to opposite values, the env variable will be the first choice?

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/openshift/odo/pkg/log"
+	"github.com/openshift/odo/pkg/preference"
+)
+
+// env variables
+const (
+	// Setting this env to `docker` will enable pushing to docker containers
+	// and will override the setting in the preferences file.
+
+	OdoPushTarget = "ODO_PUSHTARGET"
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget.go]: maybe `ODO_PUSH_TARGET`?

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	// Check "ODO_PUSHTARGET" env variable.
+	pushTargetEnv, _ = os.LookupEnv(OdoPushTarget)
+	if pushTargetEnv == "docker" || pushTargetEnv == "kube" {
+		return pushTargetEnv == "docker"
+	}
+
+	// Fetch odo preferences and check if pushtarget is set
+	cfg, err := preference.New()
+	if err != nil {
+		log.Errorf("failed to read odo preferences config. err: '%v'\n", err)
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget.go]: maybe use glog here and also return false after logging. Then we won't need the else block and the variable pushTargetPreference.

File: pkg/odo/util/pushtarget/pushtarget_test.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/openshift/odo/pkg/preference"
+)
+
+func TestIsDockerPushTargetEnabled(t *testing.T) {
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget_test.go]: Please change the name to the corresponding function it tests `PushTargetDocker`.

File: pkg/odo/util/pushtarget/pushtarget_test.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+	const (
+		pushTargetSetting string = "PushTarget"
+	)
+
+	// temp config file
+	tempConfigFile, err := ioutil.TempFile("", "odoconfig")
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget_test.go]: maybe use the fake filesystem here?

File: pkg/odo/util/pushtarget/pushtarget_test.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+			setConfig: "kube",
+			want:      true,
+		},
+		{
+			name:      "override pushtarget prefrence with kube pushtarget env",
+			setEnv:    "kube",
+			setConfig: "docker",
+			want:      false,
+		},
+	}
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget_test.go]: please add a case where the env is neither `kube` nor `docker`

File: pkg/odo/util/pushtarget/pushtarget_test.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			if tt.setEnv != "" {
+				err := os.Setenv(OdoPushTarget, tt.setEnv)
+				if err != nil {
+					t.Errorf("failed to set env %s. err: '%v'", OdoPushTarget, err)
+				}
+				defer os.Unsetenv(OdoPushTarget)
+			}
+
+			// create new preference file
+			cfg, err := preference.NewPreferenceInfo()
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget_test.go]: maybe use the fake filesystem here too?

File: pkg/preference/preference.go [hunk 8]
```diff
@@ -189,7 +199,7 @@ func (c *PreferenceInfo) SetConfiguration(parameter string, value string) error
 				return errors.Wrapf(err, "unable to set %s to %s", parameter, value)
 			}
 			if typedval < 0 {
-				return errors.Errorf("cannot set timeout to less than 0")
```
Comment (mik-dass) [inline pkg/preference/preference.go]: why was this line removed?

File: pkg/preference/preference.go [hunk 9]
```diff
@@ -294,6 +310,16 @@ func (c *PreferenceInfo) GetExperimental() bool {
 	return *c.OdoSettings.Experimental
 }
 
+// GetPushTarget returns the value of PushTarget from config
```
Comment (mik-dass) [inline pkg/preference/preference.go]: `from preference`

File: pkg/odo/util/pushtarget/pushtarget_test.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+	// test table
+	tests := []struct {
+		name      string
+		setEnv    string
+		setConfig string
+		want      bool
+	}{
+		{
+			name:      "no pushtarget setting set",
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget_test.go]: maybe add some case numbering?

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+// env variables
+const (
+	// Setting this env to `docker` will enable pushing to docker containers
+	// and will override the setting in the preferences file.
+
+	OdoPushTarget = "ODO_PUSHTARGET"
+)
+
+// IsPushTargetDocker checks if the push target preference has been set to docker
+func IsPushTargetDocker() bool {
```
Comment (mik-dass) [inline pkg/odo/util/pushtarget/pushtarget.go]: Maybe pass along a filesystem. That way we can use the filesystem in our tests and speed it up.

**Comment by mik-dass** [comment]
Works well :+1:  Just added some suggestions and fixes.

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	//		- setting odo preference using "odo preference set pushtarget docker"
+
+	var (
+		pushTargetEnv        string
+		pushTargetPreference string
+	)
+
+	// Check "ODO_PUSHTARGET" env variable.
+	pushTargetEnv, _ = os.LookupEnv(OdoPushTarget)
+	if pushTargetEnv == "docker" || pushTargetEnv == "kube" {
```
Comment (kadel) [inline pkg/odo/util/pushtarget/pushtarget.go]: If I set `ODO_PUSHTARGET=foo" than it will get silently ignored and value will be taken from the preference file. 
Wouldn't it be better to fail if there is a nonsense value?

File: pkg/preference/preference.go (hunk 8 above)
Comment (kadel) [reply pkg/preference/preference.go]: even linter is complaining ;-)
```
golangci-lint run ./... --timeout 5m
pkg/preference/preference.go:201:4: SA9003: empty branch (staticcheck)
			if typedval < 0 {
			^
make: *** [golint] Error 1
```

**Comment by mik-dass** [comment]
Also @johnmcollier Please add some integration tests

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 11 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: @mik-dass How would the filesystem be passed in? `preference.New()` (which `IsPushTargetDocker` calls) doesn't take in any args, so the preferences package would need to be refactored quite a bit to support `filesystem.Filesystem`

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 1 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: Yeah, my thinking was to have the environment var override what was set in the preferences file (perhaps the user wants to use `ODO_PUSH_TARGET=docker` for a one off `odo push`, without wanting to change the global preferences.

File: pkg/preference/preference.go (hunk 8 above)
Comment (johnmcollier) [reply pkg/preference/preference.go]: Yikes, that's embarrassing, I'm not sure how I missed that or how it happened. But I've fixed it either way.

File: pkg/preference/preference.go (hunk 9 above)
Comment (johnmcollier) [reply pkg/preference/preference.go]: Good catch! I've updated the other functions from preferences that were referencing config as well.

File: pkg/odo/util/pushtarget/pushtarget_test.go (hunk 10 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget_test.go]: Good idea, done!

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 12 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: That seems like a reasonable approach 👍 , I've updated the code to log an error and erturb false if the environment var has a nonsense value (and included a new test to catch that)

**Comment by johnmcollier** [comment]
Thanks for reviewing @mik-dass @kadel! I've updated the PR with your requested changes and integration tests.

The following integration tests were added:
- Set pushtarget to docker
- Set pushtarget to Kube
- Unset pushtarget
- Setting pushtarget to invalid value

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 2 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: Good idea, updated.

File: pkg/odo/util/pushtarget/pushtarget_test.go (hunk 6 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget_test.go]: Done, updated!

File: pkg/odo/util/pushtarget/pushtarget_test.go (hunk 4 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget_test.go]: Updated.

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 3 above)
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: Good idea, I used glog for both of the errors, and made sure to return false in both cases.

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 11 above)
Comment (mik-dass) [reply pkg/odo/util/pushtarget/pushtarget.go]: Yeah you're right. Sorry about this, I didn't realize we don't have support of fake filesystems for preferences.

File: pkg/odo/util/pushtarget/pushtarget.go (hunk 12 above)
Comment (mik-dass) [reply pkg/odo/util/pushtarget/pushtarget.go]: > Wouldn't it be better to fail if there is a nonsense value?

Or maybe continue and look into the preference file instead of failing?

**Comment by johnmcollier** [comment]
More flakiness, different test fails each time 😢 

/test v4.1-integration-e2e-benchmark

**Comment by johnmcollier** [comment]
/test v4.1-integration-e2e-benchmark 

8th time's the charm?

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	// Fetch odo preferences and check if pushtarget is set
+	cfg, err := preference.New()
+	if err != nil {
+		glog.Errorf("failed to read odo preferences config. err: '%v'\n", err)
+		return false
+	}
+
+	return cfg.GetPushTarget() == "docker"
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: I would say constants are fine, but I don't think we need a custom type for them

File: pkg/odo/util/pushtarget/pushtarget.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+// env variables
+const (
+	// Setting this env to `docker` will enable pushing to docker containers
+	// and will override the setting in the preferences file.
+
+	OdoPushTarget = "ODO_PUSH_TARGET"
+)
+
+// IsPushTargetDocker checks if the push target preference has been set to docker
+func IsPushTargetDocker() bool {
```
[the remark this answers is not in the document]
Comment (johnmcollier) [reply pkg/odo/util/pushtarget/pushtarget.go]: IMO, we should keep it as is, and add other functions to the pushtarget package as needed later.

[the remark this answers is not in the document]
Comment (rajivnathan) [reply pkg/odo/util/pushtarget/pushtarget.go]: Ok, I was more just thinking about it from an adapter perspective. This way you'd call IsPushTargetDocker() to decide between docker/kube adapter. With getPushTarget() you'd be able to get it and then switch based on the type which I think is cleaner but you're right doesn't matter much right now.
